File: apps/quickink/android/app/src/main/java/app/quickink/mobile/features/scan/VoiceNoteCapturePane.kt
/*
 * VoiceNoteCapturePane.kt
 *
 * Pre-review surface that lets the user dictate a quick voice note
 * right after a fresh scan and before the full review screen. The
 * recorded clip is persisted into `voice_notes` against the in-
 * flight capture so the transcript can feed the AI-suggested tags
 * strip on the review screen (alongside the first-page OCR text).
 *
 * Layout: slim header with Skip, the existing
 * [VoicePageRecorder] body, and a bottom bar that toggles between:
 *   - Idle: single "Continue to review" CTA that skips the voice
 *     note entirely. Skip lives in the header for the same effect.
 *   - Recording: a single full-width "Cancel" button that drops the
 *     in-progress clip and returns to idle on the same pane. There
 *     is no inline save button — the recorder's own stop control
 *     ends the take, fires `onSave`, and the parent surface advances
 *     to the transcript-edit page.
 *
 * `onContinue` carries the just-recorded voice-note id (or null for
 * Skip / idle-Continue). The parent [ScanCaptureSurface] uses that
 * to decide whether to mount [VoiceNoteTranscriptionPane] next or
 * jump straight to [ScanReviewScreen].
 */

package app.quickink.mobile.features.scan

import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.WindowInsets
import androidx.compose.foundation.layout.asPaddingValues
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.statusBars
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.rememberCoroutineScope
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import app.quickink.mobile.QuickInkApp
import app.quickink.mobile.data.voicenote.SpeechTranscriber
import app.quickink.mobile.data.voicenote.TranscribeResult
import app.quickink.mobile.data.voicenote.VoiceNoteRepository
import app.quickink.mobile.ui.theme.LocalQuickInkColors
import app.quickink.mobile.ui.theme.LocalQuickInkTypography
import app.quickink.mobile.ui.theme.QuickInkSpacing
import app.releaf.mobile.ui.theme.AppSpacing
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.launch

@Composable
fun VoiceNoteCapturePane(
    captureId: String,
    userId: String,
    onContinue: (voiceNoteId: String?) -> Unit,
    autoSkipExistingNote: Boolean = true,
) {
    val colors  = LocalQuickInkColors.current
    val type    = LocalQuickInkTypography.current
    val context = LocalContext.current
    val app     = context.applicationContext as QuickInkApp
    val repo    = remember(app) { VoiceNoteRepository(app.database.voiceNoteDao()) }
    val scope   = rememberCoroutineScope()
    // Long-lived scope for the fire-and-forget transcription work
    // so it isn't cancelled when this pane unmounts on advance to
    // the transcript-edit pane. The SupervisorJob isn't auto-
    // cancelled on disposal — jobs launched on it run to completion.
    val backgroundScope = remember {
        CoroutineScope(SupervisorJob() + Dispatchers.IO)
    }

    // Auto-skip guard. On first mount we query the repository
    // for any existing voice note attached to this captureId; if
    // one is already there (e.g. video capture pre-attached the
    // extracted audio), advance to the transcript-review pane
    // with that note's id so the user can review/edit it before
    // metadata review. `checkComplete` flips to true once the
    // query lands; while it's false we render nothing (just the
    // warm bg) so the recorder UI doesn't flash for a frame.
    var checkComplete by remember { mutableStateOf(false) }
    LaunchedEffect(captureId, autoSkipExistingNote) {
        val existing = if (autoSkipExistingNote) {
            runCatching { repo.firstForCapture(captureId) }.getOrNull()
        } else {
            null
        }
        if (existing != null) {
            onContinue(existing.id)
        } else {
            checkComplete = true
        }
    }
    if (!checkComplete) {
        Box(
            modifier         = Modifier.fillMaxSize().background(colors.bg),
            contentAlignment = Alignment.Center,
        ) {
            // Empty surface — the query is cheap; flashing a
            // spinner would be noisier than a brief blank frame.
        }
        return
    }

    // Engine is owned here so the bottom bar can observe
    // `isRecording` and drive stop/cancel.
    val engine = remember { VoicePageRecorderEngine(context, scope) }
    val isRecording = engine.isRecording

    val statusBarTop = WindowInsets.statusBars.asPaddingValues().calculateTopPadding()

    Column(
        modifier = Modifier
            .fillMaxSize()
            .background(colors.bg)
            .padding(top = statusBarTop),
    ) {
        // Header
        Row(
            modifier = Modifier
                .fillMaxWidth()
                .padding(
                    start  = QuickInkSpacing.s4,
                    end    = QuickInkSpacing.s4,
                    top    = QuickInkSpacing.s8,
                    bottom = QuickInkSpacing.s2,
                ),
            verticalAlignment = Alignment.CenterVertically,
        ) {
            Spacer(modifier = Modifier.size(50.dp))
            Spacer(modifier = Modifier.weight(1f))
            Column(horizontalAlignment = Alignment.CenterHorizontally) {
                Text(
                    text  = "VOICE NOTE",
                    style = type.label.copy(
                        fontSize    = 10.sp,
                        fontWeight  = FontWeight.Bold,
                        letterSpacing = 1.2.sp,
                    ),
                    color = colors.muted,
                )
                Text(
                    text  = "Add some context",
                    style = type.body.copy(
                        fontSize   = 15.sp,
                        fontWeight = FontWeight.SemiBold,
                    ),
                    color = colors.ink,
                )
            }
            Spacer(modifier = Modifier.weight(1f))
            // Hide Skip while recording — the bottom-bar Cancel is
            // the only action available mid-take.
            if (!isRecording) {
                Text(
                    text  = "Skip",
                    style = type.label.copy(
                        fontSize   = 13.sp,
                        fontWeight = FontWeight.SemiBold,
                    ),
                    color = colors.accentDeep,
                    modifier = Modifier
                        .clip(RoundedCornerShape(999.dp))
                        .clickable { onContinue(null) }
                        .padding(horizontal = 10.dp, vertical = 4.dp),
                )
            } else {
                Spacer(modifier = Modifier.size(50.dp))
            }
        }

        Spacer(modifier = Modifier.weight(1f))

        VoicePageRecorder(
            onSave = { clip ->
                scope.launch {
                    val row = runCatching {
                        repo.insert(
                            captureId  = captureId,
                            userId     = userId,
                            audioUri   = clip.uri,
                            durationMs = clip.durationMs,
                        )
                    }.getOrNull()
                    if (row != null) {
                        // Hand the transcription off to the process-
                        // lifecycle scope so it isn't cancelled when
                        // this composable unmounts as the parent
                        // advances to the transcript-edit pane.
                        val pendingUserId = userId
                        backgroundScope.launch(Dispatchers.IO) {
                            val result = runCatching {
                                SpeechTranscriber.transcribe(
                                    context = context,
                                    fileUri = clip.uri,
                                    userId  = pendingUserId,
                                )
                            }.getOrNull()
                            if (result is TranscribeResult.Success) {
                                runCatching {
                                    repo.setTranscription(row.id, result.text, result.source)
                                }
                            }
                        }
                    }
                    onContinue(row?.id)
                }
            },
            onCancel = { /* recorder cancelled — stay on the pane */ },
            modifier = Modifier.padding(horizontal = QuickInkSpacing.s4),
            engine = engine,
        )

        Spacer(modifier = Modifier.weight(1f))

        // Bottom bar
        Box(
            modifier = Modifier
                .fillMaxWidth()
                .padding(
                    start  = AppSpacing.s5,
                    end    = AppSpacing.s5,
                    bottom = AppSpacing.s5,
                ),
        ) {
            if (isRecording) {
                // Single full-width Cancel — drops the in-progress
                // clip and returns to idle. The recorder's own stop
                // pill is the way to commit the take and advance.
                Box(
                    modifier = Modifier
                        .fillMaxWidth()
                        .clip(RoundedCornerShape(999.dp))
                        .background(Color.White.copy(alpha = 0.85f))
                        .border(
                            width = 1.dp,
                            color = colors.border,
                            shape = RoundedCornerShape(999.dp),
                        )
                        .clickable { engine.cancel() }
                        .padding(vertical = AppSpacing.s3),
                    contentAlignment = Alignment.Center,
                ) {
                    Text(
                        text  = "Cancel",
                        style = type.body.copy(fontWeight = FontWeight.SemiBold),
                        color = colors.ink,
                    )
                }
            } else {
                Box(
                    modifier = Modifier
                        .fillMaxWidth()
                        .clip(RoundedCornerShape(999.dp))
                        .background(colors.accent)
                        .clickable { onContinue(null) }
                        .padding(vertical = AppSpacing.s3),
                    contentAlignment = Alignment.Center,
                ) {
                    Text(
                        text  = "Continue to review",
                        style = type.body.copy(fontWeight = FontWeight.SemiBold),
                        color = colors.textOnAccent,
                    )
                }
            }
        }
    }
}

File: apps/quickink/ios/QuickInk/Scan/VoiceNoteCapturePane.swift
/*
 * VoiceNoteCapturePane.swift
 *
 * Pre-review surface that lets the user dictate a quick voice note
 * right after a fresh scan and before the full review screen. The
 * recorded clip is persisted into `voice_notes` against the in-
 * flight capture so the transcript can feed the AI-suggested tags
 * strip on the review screen (alongside the first-page OCR text).
 *
 * Layout: a slim header with Skip, the existing
 * `VoicePageRecorder` body, and a bottom bar that toggles between
 * two states:
 *
 *   - Idle (not recording): single "Continue to review" CTA. The
 *     user can also skip from the header.
 *   - Recording: two side-by-side buttons — "Cancel" (drops the
 *     in-progress clip and returns to idle on the same pane) and
 *     "Save and continue" (stops the recorder, commits the clip,
 *     kicks off background transcription, advances to review).
 */

import SwiftUI
import ReleafCoreDesignSystem

struct VoiceNoteCapturePane: View {

    let captureId: String
    let userId: String
    let onContinue: (String?) -> Void
    let autoSkipExistingNote: Bool

    @StateObject private var model: VoiceNoteCapturePaneModel
    @StateObject private var engine = VoicePageRecorderEngine()
    /// Auto-skip guard. On first mount we query the repository
    /// for any existing voice note attached to the captureId; if
    /// one is already there (e.g. video capture pre-attached the
    /// extracted audio), pass that note id up to transcript review
    /// so the user isn't prompted to record over audio we already
    /// have. `checkComplete` flips to true once the query lands;
    /// `body` shows a brief progress view in the meantime so we
    /// don't flash the recorder for a frame.
    @State private var checkComplete: Bool = false

    init(
        captureId: String,
        userId: String,
        onContinue: @escaping (String?) -> Void,
        autoSkipExistingNote: Bool = true
    ) {
        self.captureId = captureId
        self.userId    = userId
        self.onContinue = onContinue
        self.autoSkipExistingNote = autoSkipExistingNote
        _model = StateObject(
            wrappedValue: VoiceNoteCapturePaneModel(
                captureId: captureId,
                userId:    userId
            )
        )
    }

    var body: some View {
        if !checkComplete {
            preCheckLoader
                .task {
                    await runExistenceCheck()
                }
        } else {
            mainPane
        }
    }

    @ViewBuilder
    private var preCheckLoader: some View {
        ZStack {
            QuickInkColors.bg.ignoresSafeArea()
            ProgressView()
                .progressViewStyle(.circular)
                .tint(QuickInkColors.accent)
        }
    }

    /// Hit the voice-note repository once on mount. If a row
    /// already exists for this captureId, skip the recorder and
    /// pass that row id up so the parent can show transcript
    /// review before metadata review.
    /// Failures (DB read error) fall through to showing the pane
    /// so the user isn't stranded.
    private func runExistenceCheck() async {
        let existing = autoSkipExistingNote
            ? (try? await VoiceNoteRepository().firstForCapture(captureId))
            : nil
        if let existing {
            await MainActor.run { onContinue(existing.id) }
        } else {
            await MainActor.run { checkComplete = true }
        }
    }

    @ViewBuilder
    private var mainPane: some View {
        VStack(spacing: 0) {
            header
                .padding(.horizontal, QuickInkSpacing.s4)
                .padding(.top, QuickInkSpacing.s8)

            Spacer(minLength: QuickInkSpacing.s5)

            VoicePageRecorder(
                engine: engine,
                onSave: { clip in
                    // Reached when the recorder's internal stop fires
                    // (auto-stop at 2:00 or in-recorder Stop tap). The
                    // pane's "Save and continue" path goes through
                    // `saveAndContinue` directly, but the in-recorder
                    // stop should land the same place.
                    Task {
                        let noteId = await model.commit(uri: clip.uri, durationMs: clip.durationMs)
                        await MainActor.run { onContinue(noteId) }
                    }
                },
                onCancel: { /* recorder cancelled — stay on the pane */ }
            )
            .padding(.horizontal, QuickInkSpacing.s4)

            Spacer(minLength: QuickInkSpacing.s5)

            bottomBar
                .padding(.horizontal, QuickInkSpacing.s5)
                .padding(.bottom, QuickInkSpacing.s5)
        }
        .frame(maxWidth: .infinity, maxHeight: .infinity)
        .background(QuickInkColors.bg.ignoresSafeArea())
    }

    // MARK: Header

    private var header: some View {
        HStack(alignment: .center) {
            Color.clear.frame(width: 50, height: 1)

            Spacer()

            VStack(spacing: 2) {
                Text("VOICE NOTE")
                    .font(.system(size: 10, weight: .bold))
                    .tracking(1.2)
                    .foregroundStyle(QuickInkColors.muted)
                Text("Add some context")
                    .font(.system(size: 15, weight: .semibold))
                    .foregroundStyle(QuickInkColors.ink)
            }

            Spacer()

            if engine.isRecording {
                // Hide the Skip control while recording — Cancel /
                // Save and continue at the bottom take over.
                Color.clear.frame(width: 50, height: 1)
            } else {
                Button(action: { onContinue(nil) }) {
                    Text("Skip")
                        .font(.system(size: 13, weight: .semibold))
                        .foregroundStyle(QuickInkColors.accentDeep)
                }
                .buttonStyle(.plain)
                .accessibilityLabel("Skip voice note")
            }
        }
    }

    // MARK: Bottom bar

    @ViewBuilder
    private var bottomBar: some View {
        if engine.isRecording {
            HStack(spacing: QuickInkSpacing.s3) {
                Button(action: cancelRecording) {
                    Text("Cancel")
                        .font(AppText.body)
                        .foregroundStyle(QuickInkColors.ink)
                        .frame(maxWidth: .infinity)
                        .padding(.vertical, AppSpacing.s3)
                        .background(
                            Capsule().fill(Color.white.opacity(0.85))
                        )
                        .overlay(
                            Capsule().stroke(QuickInkColors.border, lineWidth: 1)
                        )
                }
                .buttonStyle(.plain)

                Button(action: saveAndContinue) {
                    Text("Save and continue")
                        .font(AppText.body)
                        .foregroundStyle(QuickInkColors.textOnAccent)
                        .frame(maxWidth: .infinity)
                        .padding(.vertical, AppSpacing.s3)
                        .background(QuickInkColors.accent)
                        .clipShape(Capsule())
                }
                .buttonStyle(.plain)
            }
        } else {
            Button(action: { onContinue(nil) }) {
                Text("Continue to review")
                    .font(AppText.body)
                    .foregroundStyle(QuickInkColors.textOnAccent)
                    .frame(maxWidth: .infinity)
                    .padding(.vertical, AppSpacing.s3)
                    .background(QuickInkColors.accent)
                    .clipShape(Capsule())
            }
            .buttonStyle(.plain)
        }
    }

    // MARK: Actions

    /// Stop the recorder, commit the resulting clip, kick off the
    /// background transcription, then advance to review.
    private func saveAndContinue() {
        guard let result = engine.stop() else {
            // Recording was below the engine's min duration — drop
            // it and advance anyway so the user isn't trapped.
            onContinue(nil)
            return
        }
        Task {
            let noteId = await model.commit(uri: result.uri, durationMs: result.durationMs)
            await MainActor.run { onContinue(noteId) }
        }
    }

    /// Drop the in-progress clip and return to the pane's idle
    /// state. User can tap to record again or skip.
    private func cancelRecording() {
        engine.cancel()
    }
}

@MainActor
final class VoiceNoteCapturePaneModel: ObservableObject {

    let captureId: String
    let userId: String
    private let repo: VoiceNoteRepository

    init(captureId: String, userId: String, repo: VoiceNoteRepository = VoiceNoteRepository()) {
        self.captureId = captureId
        self.userId    = userId
        self.repo      = repo
    }

    /// Insert the freshly recorded clip and fire-and-forget a
    /// transcription pass. We don't block the UI on transcription —
    /// the review screen reactively pulls in the transcript once
    /// `setTranscription` lands, and re-runs the AI suggester at
    /// that point.
    func commit(uri: String, durationMs: Int) async -> String? {
        do {
            let row = try await repo.insert(
                captureId:  captureId,
                userId:     userId,
                audioUri:   uri,
                durationMs: durationMs
            )
            let pendingUserId = userId
            let pendingCaptureId = captureId
            Task.detached(priority: .userInitiated) {
                let granted = await VoiceTranscriber.requestPermission()
                guard granted else { return }
                guard let url = Self.fileURL(for: uri) else { return }
                guard let result = await VoiceTranscriber.transcribe(
                    fileURL: url,
                    userId:  pendingUserId
                ) else { return }
                try? await VoiceNoteRepository().setTranscription(
                    id:            row.id,
                    transcription: result.text,
                    source:        result.source
                )
                try? await CaptureRepository().appendNote(
                    captureId: pendingCaptureId,
                    text:      result.text
                )
            }
            return row.id
        } catch {
            print("VoiceNoteCapturePane commit failed: \(error)")
            return nil
        }
    }

    nonisolated private static func fileURL(for raw: String) -> URL? {
        if let parsed = URL(string: raw), parsed.isFileURL { return parsed }
        return URL(fileURLWithPath: raw)
    }
}
